src/sage/schemes/elliptic_curves/ell_finite_field.py

```diff
@@ -2772,17 +2852,324 @@ def find_q(m, m4_fac, D):
+            EllipticCurve(GF(3), [0, 2, 0, 0, 2]),
+            EllipticCurve(GF(5), [2, 0])
+        ]
+        return
+
+    # We start with small primes directly to accelerate the search. Note that
+    # 1000 is a magic constant, it's just fast enough to compute without
+    # sacrificing much speed.
+    # The if-then-else term is (-1)^((p - 1) / 2) * p in [BS2007]_ page 5.
+    S = [(-p if p % 4 == 3 else p) for p in prime_range(3, min(1000, 4 * N))
+         if legendre_symbol(N, p) == 1]
+
+    def abs_products_under(bound):
+        """
+        This function returns an iterator of all numbers with absolute value not
+        exceeding ``bound`` expressable as product of distinct elements in ``S``
+        in ascending order.
+        """
+        import heapq
+        hq = [(1, 1, -1)]
+        while len(hq):
+            abs_n, n, idx = heapq.heappop(hq)
+            yield n
+            for nxt in range(idx + 1, len(S)):
+                if abs_n * abs(S[nxt]) <= bound:
+                    heapq.heappush(hq, (abs_n * abs(S[nxt]), n * S[nxt], nxt))
+                else:
+                    break
+
+    # We add p = 1 to process the small primes.
+    for p in itertools.chain([1], Primes()):
+        if p != 1:
+            if p < abs(S[-1]):
+                continue
+
+            if legendre_symbol(N, p) != 1:
+                continue
+
+            # Later we need x^2 + (-D)y^2 = 4N, and since y = 0 has no
+            # solution, we need p = |p_star| <= |-D| <= 4N. This is a stopping
+            # condition for the algorithm.
+            if p > 4 * N:
+                break
+
+            verbose(f"Considering {len(S) + 1}th valid prime {p}", level=3)
+
+        p_star = -p if p % 4 == 3 else p
+
+        for e in abs_products_under(4 * N // p):
+            # According to the paper, the expected minimum D to work is
+            # O(log(N)^2)
+            D = p_star * e
+            assert abs(D) <= 4 * N
+
+            if D % 8 != 5 or D >= 0:
+                continue
+
+            verbose(f"Testing {D=}", level=4)
+
+            Q = BinaryQF([1, 0, -D])
+            sol = Q.solve_integer(4 * N, algorithm='cornacchia')
+            if sol is None:
+                continue
+
+            x, _ = sol
+            for p_i in [N + 1 - x, N + 1 + x]:
+                if is_prime(p_i):
+                    verbose(f"Computing the Hilbert class polynomial H_{D}",
+                            level=2)
+                    H = hilbert_class_polynomial(D)
+                    K = GF(p_i)
+                    for j0 in H.roots(ring=K, multiplicities=False):
+                        E = EllipticCurve(K, j=j0)
+                        # `E.twists()` also contains E.
+                        for Et in E.twists():
+                            # `num_checks=1` is sufficient for prime order.
+                            if Et.has_order(N, num_checks=1):
+                                Et.set_order(N, check=False)
```

Comment:
Do you think it's worth calling `set_order` in `has_order`, at least when you're in the situation where you know `has_order` will return the correct result?

Reply:
Yeah we can do that, I was worried because there's a bug report open right now, but I think it'll be fine...